binder/src/main/java/io/grpc/binder/internal/ActiveTransportTracker.java

```diff
@@ -0,0 +1,95 @@
+package io.grpc.binder.internal;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import javax.annotation.concurrent.GuardedBy;
+import io.grpc.Attributes;
+import io.grpc.Metadata;
+import io.grpc.internal.ServerListener;
+import io.grpc.internal.ServerStream;
+import io.grpc.internal.ServerTransport;
+import io.grpc.internal.ServerTransportListener;
+
+/**
+ * Tracks which {@link BinderTransport.BinderServerTransport} are currently active and allows
+ * invoking a {@link Runnable} only once all transports are terminated.
+ */
+final class ActiveTransportTracker implements ServerListener  {
+  private final ServerListener delegate;
+  private final Runnable terminationListener;
+  @GuardedBy("this")
+  private boolean shutdown = false;
+  @GuardedBy("this")
+  private int activeTransportCount = 0;
+
+  /**
+   * @param terminationListener invoked only once the server has started shutdown
+   * ({@link #serverShutdown()} AND the last active transport is terminated.
+   */
+  ActiveTransportTracker(ServerListener delegate, Runnable terminationListener) {
+    this.delegate = delegate;
+    this.terminationListener = terminationListener;
+  }
+
+  @Override
+  public ServerTransportListener transportCreated(ServerTransport transport) {
+    checkState(
+        !shutdown,
+        "Attempting to track a new BinderServerTransport, but the server shutdown has already" +
+            " started");
+    activeTransportCount++;
+    ServerTransportListener originalListener = delegate.transportCreated(transport);
+    return new TrackedTransportListener(originalListener);
+  }
+
+  private void untrack() {
+    synchronized (this) {
+      activeTransportCount--;
+      if (shutdown && activeTransportCount == 0) {
+        terminationListener.run();
+      }
+    }
+  }
+
+  @Override
+  public void serverShutdown() {
+    synchronized (this) {
+      shutdown = true;
+
+      if (activeTransportCount == 0) {
+        // We can run immediately
+        terminationListener.run();
```

Comment:
Consider invoking this callback after releasing the synchronized lock, just in case

Reply:
You may want to factor this new more elaborate logic out into a method like `maybeInvokeTerminationCallback()`

Reply:
Probably doesn't matter since this class shouldn't be interacted with after the termination callback is invoked, but agreed in principle that it may be a good idea to release locks earlier if possible.

Updated.

Reply:
Thanks but it now looks like every call to maybeInvokeTerminationCallback() comes from an already synchronized block so I don't think anything has changed...

For me, the important principle is that, in general, holding a lock while invoking an arbitrary user callback risks deadlock. One DRY approach could be:
```
@GuardedBy("this")
private Runnable getListenerIfTerminated() {
  return (shutdown && activeTransportCount == 0) ? terminationListener : null;
}

private void untrack() {
  Runnable runMeUnlocked;    
  synchronized (this) {
    --activeTransportCount;
    runMeUnlocked = getListenerIfTerminated();
  }
  if (runMeUnlocked != null) { runMeUnlocked.run(); }
}

@Override
public void serverShutdown() {
  delegate.serverShutdown();
  Runnable runMeUnlocked;    
  synchronized (this) {
    shutdown = true;
    runMeUnlocked = getListenerIfTerminated();
  }
  if (runMeUnlocked != null) { runMeUnlocked.run(); }
}
```

OTOH you have a point that no calls after termination are allowed, per ServerListener contract, so feel free to make that case in a comment instead.

Reply:
Reworked it based on your suggestion. Even if the contract should be stable, holding a lock while some unknown code runs feels like something to avoid regardless.